src/CollectionCrawlQueue.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+namespace Spatie\Crawler;
+
+use Illuminate\Support\Collection;
+use Spatie\Crawler\Exception\UrlNotFoundByIndex;
+
+class CollectionCrawlQueue implements CrawlQueue
+{
+    /** @var \Illuminate\Support\Collection */
+    public $urls;
+
+    /** @var \Illuminate\Support\Collection */
+    public $pendingUrls;
+
+    public function __construct()
+    {
+        $this->urls = collect();
+
+        $this->pendingUrls = collect();
+    }
+
+    public function add(CrawlUrl $url): CrawlQueue
+    {
+        if ($this->has($url)) {
+            return $this;
+        }
+
+        $this->urls->push($url);
+
+        $url->setId($this->urls->keys()->last());
+        $this->pendingUrls->push($url);
+
+        return $this;
+    }
+
+    public function hasPendingUrls(): bool
+    {
+        return (bool) $this->pendingUrls->count();
+    }
+
+    /**
+     * @param int $id
+     *
+     * @return \Spatie\Crawler\CrawlUrl|null
+     */
+    public function getUrlById(int $id): CrawlUrl
+    {
+        if (! isset($this->urls->values()[$id])) {
+            throw new UrlNotFoundByIndex(sprintf('#%d crawl url not found in collection', $id));
```

Comment:
`sprintf` doesn't get used often in our code bases. Replace this with string interpolation.

Reply:
Sorry, fixed. 

I have removed the PDO based driver. I thought it could be a good example, but you are right.